### ISSUE TITLE
The include-filter and exclude-filter values on p:directory-list

### ISSUE THREAD

**Issue by ndw** [xproc/3.0-specification#507]
We decided to make the filters on p:directory-list a sequence of strings (regular expressions, but that’s not relevant here).

I’m not suggesting we should change that. That’s fine.

But note that it is new and slightly odd. There are no other options that have that type. And it has some possibly odd consequences.

```
<p:directory-list path="."
                  include-filter="('a.*','b.*','c.*', '.2.*')"
                  exclude-filter="'d.*'"/>
```

First, note that the single value in `exclude-filter` needs to be quoted in order to be a singleton list of strings. I think we could agree to special-case this and say that "d.*" is a singleton string, but if you employ a general purpose sequence parser, I think it’ll need to be special cased.

(What’s the rule? If, after normalizing whitespace, the value of the attribute does not begin with “(“, “'”, or ‘"’ then we treat the literal value as a string value?)

Should we special case it?

I think it’ll surprise users if we don’t, but if we do, that has the potential to be surprising as well. (You’ll have to add quotes and parens to make a list of two regular expressions.)

Second, do we cast non-string members to strings? Is this allowed:

```
include-filter="('a', b, p:foo, Q{uri}localname, 17, 3.1415)"/>
```

I propose not.

Third, passing a list of strings as an AVT isn’t going to work as expected because AVTs are strings and the string-value of a list of strings isn’t the same as the lexical representation of a list of strings. Alas. We probably just have to document that.

I wonder if something like this would work?

```
<p:variable name="list" select="('a.*','b.*','c.*', '.2.*')"/>

<p:directory-list path="."
                    include-filter="{concat('('',
                                     string-join($list, ''','''),
                                     ')')}"
                    exclude-filter="('d.*', 'c1.*')"/>
```

I can’t tell because there’s some bug in my AVT code :-/

On further consideration, another idea occurs to me: if the options are specified as shortcuts, they are treated as a single regex. So you can't do what I wrote above, you have to do this:

```xml
<p:directory-list path="." exclude-filter="d.*">
   <p:with-option name="include-filter" select="('a.*','b.*','c.*', '.2.*')"/>
</p:directory-list>
```

Now you don't have to quote the string in the exclude filter, because it's just a single regex by definition, and if you want a list, you have to use the longer `p:with-option` form.

I think that's going to be easier to explain.

Thoughts?

**Comment by gimsieke** [comment]
Yes, I’m in favor of only allowing a single regex in the attribute options.

If we change it to single strings, we should then re-consider allowing AVTs in the attribute options since this will force users to always use `{{` and `}}` instead of `{` and `}` in their regexes.

**Comment by ndw** [comment]
Putting `include-filter` and `exclude-filter` as attributes on `p:directory-list` is using the "shortcut options" syntax. I don't think we should even consider special casing that. They will be AVTs, if you want { and }, you'll have to double them. Inconvenient, perhaps, but way less inconvenient (IMHO) than another special rule.

**Comment by gimsieke** [comment]
Given that you can put the alternatives always (proof?) into a single regex, I think we can do without sequences of regexes altogether. So we don’t need them in the `<p:with-option>` filters, either.

**Comment by ndw** [comment]
*shrug*

I'm not going to fight hard in one direction or the other on this one. I think it's easier to write a sequence of simple(r) expressions than one long expression, and limiting the attribute shortcut to a single regex doesn't prevent someone from putting a complicated one in there. But I've also never wanted a very complicated expression in practice, so...

**Comment by xml-project** [comment]
I think this is fixed with pr #509 , isn't it? So could be closed.

**Comment by ndw** [comment]
Yes, I think so.
